atd-toolbox/s3_restore_cr3_pdfs.py

```diff
@@ -0,0 +1,270 @@
+            config = aws_config
+            )
+    prefix = ('production' if args.production else 'staging') + '/cris-cr3-files/'
+    s3.list_objects(Bucket = bucket, Prefix = prefix)
+except:
+    eprint("Unable to complete call to S3; check AWS credentials")
+    sys.exit(1)
+
+
+# check to see if file is available on disk
+try:
+    assert(os.path.isfile(args.crashes))
+except:
+    eprint("Crashes file is not available on disk")
+    sys.exit(1)
+
+
+# parse JSON file containing crashes
+with open(args.crashes) as input_file:
+    try:
+        crashes = json.load(input_file)['crashes']
+    except:
+        eprint("Crashes file contains invalid JSON")
+        sys.exit(1)
+
+
+# iterate over crashes found in JSON object
+for crash in crashes:
+    print("Crash: " + str(crash))
+
+    try:
+        # graphql query to get current cr3_file_metadata
+        get_metadata = """
+        query get_cr3_metadata($crashId: Int) {
+            atd_txdot_crashes(where: {crash_id: {_eq: $crashId}}) {
+                cr3_file_metadata
+                }
+            }
+        """
+
+        # get the metadata as a dict or None if null in DB
+        cr3_metadata = requests.post(HASURA_ENDPOINT, headers = HEADERS, data = json.dumps(
+            {
+            "query": get_metadata,
+            "variables": {
+                "crashId": crash
+                }
+            })).json()['data']['atd_txdot_crashes'][0]['cr3_file_metadata']
+
+    except:
+        eprint("Request to get existing CR3 metadata failed.")
+        sys.exit(1)
+
+
+    if cr3_metadata is None:
+        print("No metadata in database for crash; creating empty dict to populate")
+        cr3_metadata = {}
+
+
+    key = prefix +  str(crash) + '.pdf'
+
+    # get versions of object sorted most recent to oldest
+    versions = sorted(s3_resource.Bucket(bucket).object_versions.filter(Prefix = key), 
+                        key=attrgetter('last_modified'), reverse=True)
+
+    # keep track of if we found one for diagnostic message logging
+    previous_version_found = False
+    for version in versions:
+        obj = version.get()
+
+        # read canidate previous version into a variable
+        # this avoids needing to put the file to disk for the use of the `file` command
+        previous_version = obj['Body'].read()
+
+        # not really magic; the underlying library of the `file` command on unix is called libmagic
+        # use libmagic to figure out what kind of file the file is
+        mime_type = magic.Magic(flags = magic.MAGIC_MIME_TYPE).id_buffer(previous_version)
+        encoding = magic.Magic(flags = magic.MAGIC_MIME_ENCODING).id_buffer(previous_version)
+
+        if mime_type != 'application/pdf':
+            print("Skipping version " + obj.get('VersionId') + " because it is a " + mime_type)
+            continue
+        else:
+            print("Version " + obj.get('VersionId') + " is acceptable for restoration because it is a " + mime_type)
```

Comment:
Where the real magic is! 🧙 Would you mind setting up a requirements file @frankhereford ? so that way we know what libraries we need to install to run the code? You may be able to collect the libraries for this script using pip freeze.

```
pip freeze > requirements.txt 
```

Reply:
Yea 100%, that is a great idea. Thanks for the suggestion.

Reply:
I installed a venv, got the script running in that and then froze out the requirements with pip. Thanks again for the suggestion! I'll be sure to do any future python scripts in a venv if they are stand-alone like this one; it's so the only way to go.

Reply:
good call—and thanks @frankhereford !